tsl/src/chunk.c

```diff
@@ -1180,3 +1186,538 @@ chunk_merge_chunks(PG_FUNCTION_ARGS)
+}
+
+/*
+ * Reconstruct and rewrite the given tuple.
+ *
+ * Basically taken from heapam module.
+ *
+ * The tuple needs to be rewritten because the new tuple doesn't include,
+ * e.g., dropped columns.
+ */
+static void
+reform_and_rewrite_tuple(HeapTuple tuple, Relation splitrel, RelationSplitInfo *rsi)
+{
+	TupleDesc oldTupDesc = RelationGetDescr(splitrel);
+	TupleDesc newTupDesc = RelationGetDescr(rsi->rel);
+	HeapTuple tupcopy;
+
+	if (rsi->tupmap)
+		tupcopy = execute_attr_map_tuple(tuple, rsi->tupmap);
+	else
+	{
+		int i;
+
+		heap_deform_tuple(tuple, oldTupDesc, rsi->values, rsi->isnull);
+
+		/* Be sure to null out any dropped columns */
+		for (i = 0; i < newTupDesc->natts; i++)
+		{
+			if (TupleDescAttr(newTupDesc, i)->attisdropped)
+				rsi->isnull[i] = true;
+		}
+
+		tupcopy = heap_form_tuple(newTupDesc, rsi->values, rsi->isnull);
+	}
+
+	/* The heap rewrite module does the rest */
+	rewrite_heap_tuple(rsi->rwstate, tuple, tupcopy);
+
+	heap_freetuple(tupcopy);
+}
+
+/*
+ * Split a chunk along a given dimension and split point.
+ *
+ * The column/dimension and split_at point are optional. If these arguments
+ * are not specified, the chunk is split in two equal sizes (from a range
+ * perspective) based on the primary partitioning column.
+ */
+Datum
+chunk_split_chunk(PG_FUNCTION_ARGS)
+{
+	Oid relid = PG_ARGISNULL(0) ? InvalidOid : PG_GETARG_OID(0);
+	const char *colname = PG_ARGISNULL(1) ? NULL : PG_GETARG_CSTRING(1);
+	Relation splitrel = table_open(relid, AccessExclusiveLock);
+	bool verbose = PG_ARGISNULL(3) ? false : PG_GETARG_BOOL(3);
+	int elevel = verbose ? INFO : DEBUG1;
+
+	if (splitrel->rd_rel->relkind != RELKIND_RELATION)
+		ereport(ERROR,
+				(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+				 errmsg("cannot split non-table relations")));
+
+	/* Only owner is allowed to split */
+	if (!object_ownercheck(RelationRelationId, relid, GetUserId()))
+		aclcheck_error(ACLCHECK_NOT_OWNER,
+					   get_relkind_objtype(splitrel->rd_rel->relkind),
+					   get_rel_name(relid));
+
+	/* Lock toast table to prevent it from being concurrently vacuumed */
+	if (splitrel->rd_rel->reltoastrelid)
+		LockRelationOid(splitrel->rd_rel->reltoastrelid, AccessExclusiveLock);
+
+	/*
+	 * Check for active uses of the relation in the current transaction,
+	 * including open scans and pending AFTER trigger events.
+	 */
+	CheckTableNotInUse(splitrel, "split_chunk");
+
+	if (IsSystemRelation(splitrel))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("cannot split system catalog relations")));
+
+	Oid amoid = splitrel->rd_rel->relam;
+
+	if (amoid != HEAP_TABLE_AM_OID)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("access method \"%s\" is not supported for split", get_am_name(amoid))));
+
+	const Chunk *chunk = ts_chunk_get_by_relid(relid, true);
+
+	if (chunk->fd.compressed_chunk_id != INVALID_CHUNK_ID)
+		ereport(ERROR,
+				(errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
```

Comment:
Why this error code and not "feature not supported"? Same question in the other direction, why "feature not supported" rather than "not in required state"?

I naïvely thought that it's one or the other.

Reply:
Well, I guess both works. The thing is that you can split the chunk if you decompress. In that sense the chunk is not in a state to be split when compressed. If you feel strongly, I can change to not supported.

Reply:
Updated to use not supported error.

Reply:
I have no strong opinion on one or the other, but the error code assignment seemed arbitrary.